Fix/improve checkpoint integration tests

kolyshkin opened the pull request:
This fixes the checkpoint/restore tests which were failing or hanging.

A copy/paste of the last commit message:

Commit a9e15e7 adds a check that stdin/out/err pipes
are restored correctly. Commit ec26065 copy/pastes
the same code to one more another test.

Problem is (as pointed out in commit 5369f9a) these tests
sometimes hang. I have also seen them fail.

Apparently, the code used to create pipes and open them to fds
is racy:

```shell
cat $fifo | cat $fifo &
pid=$!
exec 50</proc/$pid/fd/0
exec 51>/proc/$pid/fd/0
```

Since `cat | cat` is spawned asynchronously, by the time exec is used,
the second cat process (i.e. $pid) is already fork'ed but it might
not be exec'ed yet. As a result, we get this (`ls -l /proc/self/fd`):

```
lr-x------. 1 root root 64 Apr 20 02:39 50 -> /dev/pts/1
l-wx------. 1 root root 64 Apr 20 02:39 51 -> /dev/pts/1
```
or, sometimes, this:
```
lr-x------. 1 root root 64 Apr 20 02:45 50 -> /dev/pts/1
l-wx------. 1 root root 64 Apr 20 02:45 51 -> 'pipe:[215791]'
```

instead of expected set of pipes:

```
> lr-x------. 1 root root 64 Apr 20 02:45 50 -> 'pipe:[215791]'
> l-wx------. 1 root root 64 Apr 20 02:45 51 -> 'pipe:[215791]'
```

One possible workaround is to add `sleep 0.1` or so after cat|cat,
but it is outright ugly (besides, we already have one sleep in
the test code).

The solution is to not use any external processes to create pipes.
I admit this still looks not very comprehensible, but at least it
is easier than before, and it works.

While at it, remove code duplication, moving the setup and check
code into a pair of functions.

Finally, since the tests are working now, remove the skip.

kolyshkin commented:
@avagin @adrianreber PTAL. Also a question -- do you guys think we need to check the pipes in two tests?

avagin commented:
LGTM, thanks!

> do you guys think we need to check the pipes in two tests?

This is our check to be sure that processes have been restored correctly.

kolyshkin commented:
Fails with cgroupv2 as it needs a fix from runc side as well as patched criu. Will add.

adrianreber commented:
@kolyshkin Is the reason to skip cgroup2 tests on vagrant-fedora 31 that this test uses CRIU from the Fedora repository (3.13) and that it does not build CRIU itself like done in the Dockerfile (3.13 + cgroup2 patches)?

adrianreber commented:
> Is the reason to skip cgroup2 tests on vagrant-fedora 31 that this test uses CRIU from the Fedora 

I meant 'Is the reason to skip checkpoint/restore tests..."

kolyshkin commented:
@adrianreber there are two reasons, both are described in commit message.

I have actually already fixed it, just want the fix to be in a separate PR (and need time to create it). Hope to do it today.

kolyshkin commented:
@adrianreber added a commit to remove the sleep from the lazy mig test, PTAL

adrianreber commented on tests/integration/checkpoint.bats:
```diff
@@ -161,8 +164,6 @@ function check_pipes() {
   cat $lazy_pipe
```
The test probably fails because this `cat` will now hang forever.

adrianreber commented:
Would be good if the sleep can be removed. @avagin and I discussed this probably one and a half year ago, but we came to the conclusion that it is complicated and did not follow up. Can't remember all the details any more. If your pipe magic can solve this I am :+1: .

kolyshkin commented:
> Would be good if the sleep can be removed. @avagin and I discussed this probably one and a half year ago, but we came to the conclusion that it is complicated and did not follow up. Can't remember all the details any more. If your pipe magic can solve this I am ￼ .

@adrianreber I've reverted the last commit for now, will address it later (or you can tackle on it).

The thing is, what advertized as
```
--status-fd value   criu writes \0 to this FD once lazy-pages is ready
```
is in fact a file not an fd:

```go
StatusFd:                context.String("status-fd"),
...
go waitForCriuLazyServer(statusRead, criuOpts.StatusFd)
...
func waitForCriuLazyServer(r *os.File, status string) error {
...
        fd, err := os.OpenFile(status, os.O_TRUNC|os.O_WRONLY, os.ModeAppend)
```

so we can not pass a file descriptor like I did in a (now reverted) commit.

kolyshkin commented:
OK this one is ready

AkihiroSuda commented:
needs rebase

kolyshkin commented:
Rebased

adrianreber commented:
As soon as CRIU 3.14 is released I will update it in Fedora 31 so you do not need to build the package from source for the vagrant-fedora-31 test environment.

kolyshkin commented:
CI failures are caused by the recent github instability issues (failure to download criu patches). Restarted CI

AkihiroSuda commented:
LGTM

thaJeztah commented:
@cyphar @mrunalp PTAL

mrunalp commented:
LGTM
